### DISCUSSION TITLE
Supporting Wayland in O3DE

### DISCUSSION THREAD

**Discussion by spham-amzn** [o3de/o3de#18185] [General]
## Overview
Opening up a discussion about how to add support for Wayland (Alternative for xcb/x11) for O3DE. When Linux supported first started for the code base, `xcb` was specified for the binary build for Qt, and from there that decision trickled down to the rest of the engine. For example, Atom uses [VkXcbSurfaceCreateInfoKHR ](https://github.com/o3de/o3de/blob/a76660992071b531dd025a573270f8d578902a9d/Gems/Atom/RHI/Vulkan/Code/Source/Platform/Linux/RHI/WSISurface_Linux.cpp#L36) to create native surface. For the core [AzFramework](https://github.com/o3de/o3de/tree/development/Code/Framework/AzFramework) library, in addition to the platform abstractions, there was an abstraction created for the UI protocols, with only [xcb](https://github.com/o3de/o3de/tree/development/Code/Framework/AzFramework/Platform/Common/Xcb)  actually implemented. 

## Existing Abstraction Support
The abstraction to be able to use different protocols is controlled by PAL traits, and currently there were 2 traits that were defined for this:

* PAL_TRAIT_LINUX_WINDOW_MANAGER_XCB
* PAL_TRAIT_LINUX_WINDOW_MANAGER_WAYLAND

These traits are set in the file [platform_nativeui_linux.cmake](https://github.com/o3de/o3de/blob/development/Code/Framework/AzFramework/Platform/Linux/platform_nativeui_linux.cmake) and controlled by [PAL_TRAIT_LINUX_WINDOW_MANAGER](https://github.com/o3de/o3de/blob/a76660992071b531dd025a573270f8d578902a9d/cmake/Platform/Linux/PAL_linux.cmake#L67-L70).

Doing a search on the traits in the source code will show you all the touch points for this abstraction.

## Issues
There are a couple of issues that exist that needs to be tackled to support wayland.

### Qt
The [3rd Party Qt package](https://github.com/o3de/3p-package-source/tree/main/package-system/Qt) for O3DE is a prebuilt binary based on a custom fork of Qt. The [build script](https://github.com/o3de/3p-package-source/blob/main/package-system/Qt/docker_build_qt_linux.sh) that builds the package is based on xcb. In order to support wayland in the editor, there will need to be a separate version of this package for Linux that is Wayland specific. That may mean the need to produce an XCB and a Wayland version of Qt.

### Wayland 
While the abstraction exists, Wayland has not been implemented or even defined. There would be need to be a bit of work to add this support. Fortunately, @DrogonMar has embarked down this path already.
 

The purpose of this discussion is to open up and get ideas and opinions for this, and any insights or gotchas, and maybe produce an RFC for this work.

**Comment by DrogonMar** [comment]
So far the GameLauncher target is ready for Wayland (still needs testing.)
My thoughts so far is that getting Qt to support Wayland would be the best as High DPI displays are getting common and X11 isn't great at handling that so while doing a work around to get only the editor to use X11 would be great its not the best way forward.

**Reply by DrogonMar** [reply]
The only thing I'm unsure about using Wayland in Qt is how to handle [NativeEventFilter.](https://github.com/o3de/o3de/blob/11e6a1e8d28035d187b0df6d0869be62c8c151aa/Code/Editor/Platform/Linux/Editor/Core/QtEditorApplication_linux.cpp#L54)
If you look at [WaylandInputDeviceMouse](https://github.com/DrogonMar/o3de/blob/wayland/Code/Framework/AzFramework/Platform/Common/Wayland/AzFramework/WaylandInputDeviceMouse.cpp) as an example you can see how libwayland handles events which is a [listener](https://github.com/DrogonMar/o3de/blob/9e454553e05de8f7daa32fc42ee36309e95de91f/Code/Framework/AzFramework/Platform/Common/Wayland/AzFramework/WaylandInputDeviceMouse.h#L110C29-L110C48) theres probably a way but this my current concern.
